### PR TITLE
Feat: Complete Neuro-Symbolic Pipeline with Docker Infra

### PR DESCRIPTION
**Pull request by NakulSingh156**
### Motivation

This PR significantly refactors the extraction pipeline, transitioning from a heuristic-based standalone script to a robust, reproducible **Neuro-Symbolic Architecture** fully aligned with the GSoC25 Redis infrastructure. The goal is to provide a deterministic, hallucination-resistant extraction layer that seamlessly integrates with DBpedia's existing backend.

### Core Architectural Updates

* **Dockerized Environment:** Orchestrated the pipeline using `docker-compose`, isolating the Python extraction environment and the Redis backend for zero-config reproducibility.
* **Redis-Backed Entity Resolution:** Replaced external API calls with an offline, high-performance Redis cache. Includes `seed_redis.py` to securely mock DBpedia fixtures (Anchor Candidates & Types) for end-to-end testing without requiring the 50GB data dump.

### Neuro-Symbolic Features

* **The "Hallucination Buster" (Ontology Guardrails):** Integrated a strict SPARQL validation layer. If the Neural model (`all-MiniLM-L6-v2`) extracts a semantically plausible but factually incorrect triple (e.g., *Ronaldo -> playsFor -> Chicago Bulls*), the Symbolic layer intercepts and rejects it as contradicting world knowledge.
* **Graph-Proximity Reasoning:** Adapted validation to handle historical/complex facts (e.g., *Ronaldo -> Real Madrid*). If strict ontology matching fails due to temporal shifts, the system falls back to a BFS traversal, exploiting `dbo:wikiPageWikiLink` to compute contextual relatedness.
* **Deterministic Extraction:** Refactored the `Spacy` pipeline to enforce deterministic Subject/Object assignment based on character-index sorting, resolving non-deterministic set behaviors.

### CI/CD & Code Quality
* Resolved all static analysis warnings flagged by **CodeRabbit AI** (removed dead code, fixed bare exceptions, sanitized return types, and standardized environment variables). 

---

### Proof of Build (End-to-End Pipeline Validation)

**1. Infrastructure: Clean Redis Seeding**
Screenshot 2026-02-14 at 9 27 58 AM

**2. Extraction & Validation: Active Voice, Passive Voice, and Hallucination Rejection**
*Notice how the symbolic layer successfully validates complex sentence structures while strictly rejecting the hallucinated 'Chicago Bulls' output.*

Screenshot 2026-02-14 at 9 27 33 AM

**3. Client-Side Graph Reasoning (BFS Traversal)**
*Finding contextual links for complex entities directly through the Knowledge Graph.*

Screenshot 2026-02-12 at 5 40 43 PM

### Things Addressed in this PR:

1. Migrated to offline Redis infrastructure per GSoC25 guidelines.
2. Corrected previous benchmark anomalies and expanded testing to a 20-item Mega-Benchmark.
3. Replaced hardcoded weights and stopwords with a dynamic HybridLinker.
4. Abstracted predicate matching via a ModelInterface using all-MiniLM-L6-v2.

### CODE REVIEW

File: GSoC26/assets/benchmark.png
Comment (mommi84) [inline GSoC26/assets/benchmark.png:1]: `AI` and `Amazon` look wrong too, so the final score should actually be 6/9, not 8/9.

File: GSoC26/src/dbpedia_linker.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
+import requests
+import re
+from rapidfuzz import fuzz
+# IMPORT YOUR RESOLVER FROM THE OTHER FILE
+from src.wiki_resolver import resolve_redirect
+
+DBPEDIA_LOOKUP = "https://lookup.dbpedia.org/api/search"
```
Comment (mommi84) [inline GSoC26/src/dbpedia_linker.py:7]: No need to use DBpedia Lookup. In this project, we indexed Wikipedia anchors into a Redis database. We query that to retrieve a list of entities given an input string.

See e.g. https://github.com/dbpedia/neural-extraction-framework/blob/main/GSoC25/EntityLinking/test_redis.py

File: GSoC26/src/dbpedia_linker.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+# Weights Configuration
+TYPE_PRIOR = {
```
Comment (mommi84) [inline GSoC26/src/dbpedia_linker.py:10]: How were these weights found?

File: GSoC26/src/dbpedia_linker.py [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    "Person": 1.8, "Athlete": 1.9, "Politician": 1.8, "Artist": 1.7,
+    "Company": 1.9, "Organisation": 1.7, "SportsTeam": 2.0,
+    "FootballClub": 2.1, "SoccerClub": 2.1, "Place": 1.7, "City": 1.7,
+    "Country": 1.7, "Album": 1.5, "Film": 1.4, "Work": 1.2,
+    "Event": 0.3, "Season": 0.05, "Tournament": 0.2
+}
+STOPWORDS = {"fc", "club", "sc", "the", "of", "and", "in", "inc"}
```
Comment (mommi84) [inline GSoC26/src/dbpedia_linker.py:17]: Why do stopwords include important words such as 'club' or 'fc'/'inc' which can distinguish, e.g., a football club from its city?

File: src/linker.py [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+import spacy
+
+# Load Spacy for NER
+try:
+    nlp = spacy.load("en_core_web_sm")
+except:
+    print("⚠️ Warning: Spacy model 'en_core_web_sm' not found. Run: python -m spacy download en_core_web_sm")
+    nlp = None
+
+WIKIPEDIA_API = "https://en.wikipedia.org/w/api.php"
```
Comment (mommi84) [inline src/linker.py]: Again, no need to use the Wikipedia API. In this project, we indexed Wikipedia redirects into a Redis database. We query that to retrieve page redirects too.

See e.g. https://github.com/dbpedia/neural-extraction-framework/blob/main/GSoC25/EntityLinking/test_redis.py

File: src/main.py [hunk 5]
```diff
@@ -0,0 +1,73 @@
+from linker import get_best_entity, extract_entity_spans
+from sentence_transformers import SentenceTransformer, util
+
+# --- 1. RELATION EXTRACTOR LOGIC (Rebel/BERT Simulation) ---
+class RelationExtractor:
+    def __init__(self):
+        self.model = SentenceTransformer('all-MiniLM-L6-v2')
+        self.predicates = {
+            "dbo:team": ["plays for", "signed with", "club", "team"],
```
Comment (mommi84) [inline src/main.py]: Why were these hard-coded?

**Review by mommi84 — CHANGES_REQUESTED** [review]
Thanks for your PR!

Please take a look at the current state of the project for the English language, which can be found in the [GSoC25 folder](https://github.com/dbpedia/neural-extraction-framework/tree/main/GSoC25).

I personally liked the idea of exploiting wiki-links to compute the relatedness between two entities in the text. However, many of the proposed features in this PR already exist in the current project.

Please go through the comments and the existing codebase before resubmitting your PR, or write a project proposal for GSoC 2026 specifying your expected roadmap to address the outstanding issues.

**Comment by NakulSingh156** [comment]
> Thanks for your PR!
> 
> Please take a look at the current state of the project for the English language, which can be found in the [GSoC25 folder](https://github.com/dbpedia/neural-extraction-framework/tree/main/GSoC25).
> 
> I personally liked the idea of exploiting wiki-links to compute the relatedness between two entities in the text. However, many of the proposed features in this PR already exist in the current project.
> 
> Please go through the comments and the existing codebase before resubmitting your PR, or write a project proposal for GSoC 2026 specifying your expected roadmap to address the outstanding issues.

Thanks a lot for the detailed review @mommi84! This is super helpful.

I’ve noted all your points and am working on the fixes:

1. Redis vs APIs: I actually missed the test_redis.py infrastructure in GSoC25. Using the internal Redis index is definitely better than calling public APIs. I’ll refactor the linker to follow that pattern.

2. Benchmarks: Good catch on 'AI' and 'Amazon'. I was being a bit too optimistic there. I’ll correct the score to 6/9.

3. Weights & Stopwords: The current weights were just based on my initial trial-and-error, I used to patch specific errors (like distinguishing "La Liga" the league vs the season). I see now that hardcoding them makes the system brittle, and removing 'FC'/'Club' was a mistake (since they distinguish the team from the city). I’ll revert to a cleaner, data-driven approach and fix the stopword list.

4. Main.py Predicates: I hardcoded those mappings solely to create a lightweight demo that runs without downloading the full 3GB extraction model. I will replace that dictionary with a proper ModelInterface so the system can load the real model (or a smaller distilled version) instead of relying on fixed strings.

I’ll push the changes shortly!

**Comment by NakulSingh156** [comment]
> Thanks for your PR!
> 
> Please take a look at the current state of the project for the English language, which can be found in the [GSoC25 folder](https://github.com/dbpedia/neural-extraction-framework/tree/main/GSoC25).
> 
> I personally liked the idea of exploiting wiki-links to compute the relatedness between two entities in the text. However, many of the proposed features in this PR already exist in the current project.
> 
> Please go through the comments and the existing codebase before resubmitting your PR, or write a project proposal for GSoC 2026 specifying your expected roadmap to address the outstanding issues.

Hi @mommi84!  Thanks again for the detailed feedback! I've just pushed a major architectural overhaul that addresses all the points:

Redis vs APIs: Fully migrated to a local Redis cache. I added a docker-compose.yml and seed_redis.py to handle the infrastructure offline, perfectly aligning with the GSoC25 test_redis.py pattern.

Benchmarks: You were right about the scoring—I've corrected it. I also expanded the testing into a 20-item "Mega-Benchmark" (documented in the README) to better track edge cases moving forward.

Weights & Stopwords: Removed the hardcoded 0.7 * fuzz logic and stopword lists in favor of a cleaner, type-aware resolution flow in the HybridLinker.

Main.py Predicates: Implemented a proper ModelInterface abstraction. The system now loads all-MiniLM-L6-v2 to calculate semantic similarity instead of relying on fixed dictionary lookups.

I've also added a Neuro-Symbolic validation layer to act as a guardrail against hallucinations. Let me know if the new architecture looks solid or if you'd like any further tweaks!

**Review by mommi84 — APPROVED** [review]
